File: agents/python-agent/hmdev/messaging/agent/core/agent_connection.py
import logging
import threading
import re
from typing import Optional, List
import uuid
import json

from hmdev.messaging.agent.core.agent_connection_event_handler import AgentConnectionEventHandler
from hmdev.messaging.agent.util import session_recovery_utility as Sess
from hmdev.messaging.agent.util.utils import sleep
from hmdev.messaging.agent.api.connection_channel_api_factory import ConnectionChannelApiFactory
from hmdev.messaging.agent.api.models import ConnectResponse, EventMessageResult, AgentInfo, ReceiveConfig
from hmdev.messaging.agent.api.impl.messaging_channel_api import MessagingChannelApi
from hmdev.messaging.agent.security.my_security import MySecurity

logger = logging.getLogger(__name__)

# Mirror Java-agent constants
DEFAULT_RECEIVE_LIMIT = 20
PASSWORD_WAIT_TIMEOUT_SECONDS = 5


class AgentConnection:

    def __init__(self, api_url: str) -> None:
        self.connection_time: Optional[float] = None
        self._agent_name: Optional[str] = None
        self._channel_api = ConnectionChannelApiFactory.get_connection_api(api_url)
        self._check_last_session = True
        self._session_id: Optional[str] = None
        self._ready_state = False
        self._channel_password: Optional[str] = None
        self._receive_thread: Optional[threading.Thread] = None
        # Private key generated when requesting a password. Kept until a PASSWORD_REPLY is received
        # or the request is cancelled/fulfilled.
        self._pending_private_key = None
        # Track the request id associated with the pending private key so replies can be correlated
        self._pending_request_id: Optional[str] = None

    # New constructor-like factory: accept optional developer_api_key
    @classmethod
    def with_api_key(cls, api_url: str, developer_api_key: Optional[str] = None):
        inst = cls(api_url)
        # Replace channel_api with one using developer_api_key
        inst._channel_api = ConnectionChannelApiFactory.get_connection_api(api_url, developer_api_key)
        return inst

    def connect(self, channel_name: str, channel_password: str, agent_name: str) -> bool:

        self._agent_name = agent_name

        if self._ready_state and self._session_id is not None:
            raise Exception(f"Agent {self._agent_name} is already connected with session {self._session_id}")

        if self._session_id is None and self._check_last_session:
            self._session_id = Sess.load_session_id(channel_name)

        resp: ConnectResponse = self._channel_api.connect(channel_name, channel_password, self._agent_name, self._session_id)

        if resp and resp.sessionId:
            try:
                # connection_time may be provided in response.date
                self._session_id = resp.sessionId
                self.connection_time = resp.date if resp.date is not None else None
            except Exception:
                # fallback: keep existing session id
                pass

            self._ready_state = True
            self._channel_password = channel_password
            if self._session_id:
                Sess.save_session_id(channel_name, self._session_id)
            logger.info("Connected to session %s", self._session_id)

            return True
        else:
            logger.warning("Connect failed: %s", resp)
            return False

    def disconnect(self) -> bool:
        if not self.is_ready():
            return False

        result: bool = self._channel_api.disconnect(self._session_id)

        if result:
            self._session_id = None
            self._ready_state = False
            # Clear any pending private key when disconnecting to avoid keeping sensitive key material alive
            self._pending_private_key = None
            self._pending_request_id = None

        return result

    def receive(self, offset_range: ReceiveConfig) -> Optional[EventMessageResult]:
        # Allow receiving messages even before the local channel password/secret is known.
        if not (self._ready_state and self._session_id):
            return None
        resp = self._channel_api.receive(self._session_id, offset_range)
        return resp

    def receive_async(self, handler: AgentConnectionEventHandler) -> None:
        if self._receive_thread is None:
            self._receive_thread = threading.Thread(target=self._run_receive, args=(handler,), daemon=True)
            self._receive_thread.start()

    def _run_receive(self, handler: AgentConnectionEventHandler) -> None:
        offset_range = ReceiveConfig(globalOffset=0, localOffset=0, limit=DEFAULT_RECEIVE_LIMIT)

        while self.is_ready():
            resp = self.receive(offset_range)
            if resp is None:
                sleep(0.5)
                continue

            events = resp.events or []
            if events:
                try:
                    handler.on_message_events(events)
                except Exception as e:
                    logger.warning("Handler raised exception: %s", e)

            if resp.nextGlobalOffset is not None:
                offset_range.globalOffset = resp.nextGlobalOffset

            if resp.nextLocalOffset is not None:
                offset_range.localOffset = resp.nextLocalOffset

            sleep(0.5)

    def send_message(self, msg: str, destination: Optional[str] = "*", as_filter_regex: Optional[bool] = True) -> bool:
        if not self.is_ready():
            return False

        # If not using regex, escape the destination (similar to Pattern.quote in Java)
        dest = destination if as_filter_regex else re.escape(destination)

        try:
            result = self._channel_api.send(msg, dest, self._session_id)
            return bool(result)
        except Exception as e:
            logger.warning("Send failed: %s", e)
            return False

    def udp_push_message(self, msg: str, destination: str) -> bool:
        """Send a message through the UDP bridge (fire-and-forget)."""
        if not self.is_ready():
            return False
        try:
            return bool(self._channel_api.udp_push(msg, destination, self._session_id))
        except Exception as e:
            logger.warning("udp_push_message failed: %s", e)
            return False

    def udp_pull(self, offset_range: ReceiveConfig) -> Optional[EventMessageResult]:
        """Pull messages through the UDP bridge with a short timeout."""
        if not self.is_ready():
            return None
        try:
            return self._channel_api.udp_pull(self._session_id, offset_range)
        except Exception as e:
            logger.warning("udp_pull failed: %s", e)
            return None

    def get_active_agents(self) -> Optional[List[AgentInfo]]:
        if not self.is_ready():
            return None

        try:
            resp = self._channel_api.get_active_agents(self._session_id)
            return resp
        except Exception as e:
            logger.warning("get_active_agents failed: %s", e)
            return None

    def set_check_last_session(self, check: bool) -> None:
        self._check_last_session = check

    def is_ready(self) -> bool:
        """Return True when the connection is ready and session id is present (matches Java isChannelReady)."""
        if not self._ready_state or self._session_id is None:
            logger.debug("Unable use channel operation, channel is not ready")
            return False
        else:
            return True

    def connect_with_channel_id(self, channel_id: str, agent_name: str, maybe_channel_name: Optional[str] = None) -> bool:
        """Connect using channelId and perform password-request flow if channel secret not available."""
        self._agent_name = agent_name
        if self._ready_state and self._session_id is not None:
            raise Exception(f"Agent {self._agent_name} is already connected with session {self._session_id}")

        if self._session_id is None and self._check_last_session:
            self._session_id = Sess.load_session_id(channel_id)

        resp: ConnectResponse = self._channel_api.connect(None, None, self._agent_name, self._session_id, channel_id)

        if resp and resp.sessionId:
            self._session_id = resp.sessionId
            self.connection_time = resp.date if resp.date is not None else None
            self._ready_state = True

            # NOTE: The automatic password-request flow has been removed from connect.
            # Call `request_password()` explicitly if needed.

            if self._session_id:
                Sess.save_session_id(channel_id, self._session_id)
            logger.info("Connected to session %s", self._session_id)
            return True
        else:
            logger.warning("Connect with channel id failed: %s", resp)
            return False

    def request_password(self, maybe_channel_name: Optional[str] = None, timeout_seconds: int = 10) -> bool:
        # Use default timeout aligned with Java agent unless caller overrides
        """Send a 'password-request' event containing a newly generated public key PEM string.

        The method will wait up to timeout_seconds for a private PASSWORD_REPLY addressed to this agent.
        The PASSWORD_REPLY content is expected to be a base64-encoded ciphertext which will be
        decrypted using the generated private key. On success the channel password is set and
        (optionally) channel secret is derived when maybe_channel_name is provided.

        Returns True if the password was received and processed, False otherwise.
        """
        if not self.is_ready() or not self._session_id:
            logger.warning("Cannot request password: connection not ready")
            return False

        if self._channel_password is not None:
            logger.debug("Channel password already known, skipping request_password")
            return True

        if not isinstance(self._channel_api, MessagingChannelApi):
            logger.warning("request_password requires MessagingChannelApi implementation")
            return False

        try:
            # use utility to generate RSA keypair and wait for reply
            from hmdev.messaging.agent.util import password_utils as PWU
            import base64, uuid, time

            priv_key, pub_pem = PWU.generate_rsa_keypair()

            # store pending private key and request id for correlation
            request_id = str(uuid.uuid4())
            self._pending_private_key = priv_key
            self._pending_request_id = request_id

            # send password-request with JSON { requestId, publicKeyPem }
            req_payload = json.dumps({"requestId": request_id, "publicKeyPem": pub_pem})
            self._channel_api.send_event('password-request', req_payload, '*', False, self._session_id)

            # wait for a PASSWORD_REPLY addressed to this agent and matching request id
            success = PWU.wait_for_password_reply(
                agent_name=self._agent_name,
                receive=lambda cfg: self.receive(cfg),
                private_key=priv_key,
                channel_api=self._channel_api,
                connection_time=self.connection_time,
                expected_request_id=request_id,
                maybe_channel_name=maybe_channel_name,
                timeout_seconds=timeout_seconds or PASSWORD_WAIT_TIMEOUT_SECONDS,
                poll_interval=0.4
            )

            # clear pending key/request id after attempt
            self._pending_private_key = None
            self._pending_request_id = None

            if success:
                # if password obtained, update local channel_password from channel_api state if set
                try:
                    # channel_api may not expose channel_password directly; keep agent's copy consistent
                    # If channel_api.channel_secret is set, we assume channel_password is derivable only if maybe_channel_name was provided
                    pass
                except Exception:
                    pass

            return bool(success)

        except Exception as e:
            logger.warning("request_password flow failed: %s", e)
            self._pending_private_key = None
            self._pending_request_id = None

        return False

File: agents/java-agent/src/main/java/com/hmdev/messaging/agent/core/AgentConnection.java
package com.hmdev.messaging.agent.core;

import com.hmdev.messaging.common.CommonUtils;
import com.hmdev.messaging.common.data.AgentInfo;
import com.hmdev.messaging.common.data.ConnectResponse;
import com.hmdev.messaging.common.data.EventMessage;
import com.hmdev.messaging.common.data.EventMessageResult;
import com.hmdev.messaging.common.data.ReceiveConfig;
import lombok.Getter;
import lombok.Setter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;


import java.util.List;
import java.util.regex.Pattern;

import com.hmdev.messaging.agent.util.Utils;

import com.hmdev.messaging.agent.api.ConnectionChannelApi;
import com.hmdev.messaging.agent.api.ConnectionChannelApiFactory;
import com.hmdev.messaging.common.crypto.EnvelopeUtil;

import java.nio.charset.StandardCharsets;
import java.security.KeyPair;
import java.time.Duration;

/**
 * Manages a messaging agent connection lifecycle: connect, send/receive (sync/async),
 * and session recovery. Business logic preserved; refactor focuses on readability
 * and maintainability while staying Java 11 compatible.
 */
public class AgentConnection {

    private static final Logger logger = LoggerFactory.getLogger(AgentConnection.class);
    private static final Long DEFAULT_RECEIVE_LIMIT = 20L ;
    private static final long PASSWORD_WAIT_TIMEOUT_SECONDS = 5;

    /**
     *  Enable/disable last-session recovery behavior at connect-time.
     */
    @Setter
    private boolean checkLastSession = true;

    private final ConnectionChannelApi channelApi;
    private String agentName;
    private String sessionId;

    private boolean readyState = false;
    private Thread receiveThread;
    @Getter
    private long connectionTime;

    /**
     * Create an AgentConnection bound to a backend API endpoint.
     *
     * @param apiUrl    Base URL of messaging API
     *
     */
    public AgentConnection(String apiUrl) {
        this(apiUrl, null);
    }

    public AgentConnection(String apiUrl, String developerApiKey) {
        this.channelApi = ConnectionChannelApiFactory.getConnectionApi(apiUrl, developerApiKey);
    }


    /**
     * Connect to a channel with given credentials.
     *
     * @param channelName     Channel to connect
     * @param channelPassword Password/secret for the channel
     * @param agentName       Name/identifier of this agent
     * @return true if connected successfully, false otherwise
     */
    public boolean connect(String channelName, String channelPassword, String agentName) throws Exception {

        this.agentName = agentName;
        if (readyState && sessionId != null) {
            throw new Exception("Agent is " + agentName + " is already connected with session " + sessionId);
        }

        if (this.sessionId == null && checkLastSession) {
            this.sessionId = Utils.loadSessionId(channelName);
        }
        ConnectResponse connectResponse = channelApi.connect(channelName, channelPassword, this.agentName, this.sessionId);

        // Success result
        if (CommonUtils.isNotEmpty(connectResponse.getSessionId())) {

            logger.debug("Connection Response is {}", connectResponse);
            this.sessionId = connectResponse.getSessionId();
            this.connectionTime =  connectResponse.getDate();
            readyState = true;
            logger.debug("Connected to session : {}", this.sessionId);

            Utils.saveSessionId(channelName, this.sessionId);
            return true;
        } else {
            return false;
        }

    }

    /**
     * Connect using a server-side channelId. Optionally provide the channelName so the agent can derive
     * the channel secret after a PASSWORD_REPLY is received (password-based secret derivation requires channelName).
     */
    public boolean connectWithChannelId(String channelId, String agentName) throws Exception {
        this.agentName = agentName;
        if (readyState && sessionId != null) {
            throw new Exception("Agent " + agentName + " is already connected with session " + sessionId);
        }

        if (this.sessionId == null && checkLastSession) {
            this.sessionId = Utils.loadSessionId(channelId);
        }

        ConnectResponse connectResponse = channelApi.connectWithChannelId(channelId, this.agentName, this.sessionId);

        if (CommonUtils.isNotEmpty(connectResponse.getSessionId())) {
            logger.debug("Connection Response is {}", connectResponse);
            this.sessionId = connectResponse.getSessionId();
            this.connectionTime = connectResponse.getDate();

            if (CommonUtils.isEmpty(channelApi.getChannelSecret())) {
                try {
                    // Generate ephemeral RSA keypair for the request
                    KeyPair keypair = EnvelopeUtil.generateRSAKeyPair();
                    String pubB64 = EnvelopeUtil.encodeRSAPublicKey(keypair.getPublic());

                    // Send PASSWORD_REQUEST event with public key
                    channelApi.send(EventMessage.EventType.PASSWORD_REQUEST, pubB64, "*" , this.sessionId, false);

                    // Poll for PASSWORD_REPLY targeted to this agent for up to 10s
                    boolean gotPassword = waitForPasswordReply(connectResponse.getMetadata().getGlobalOffset(),
                            connectResponse.getMetadata().getLocalOffset(), keypair, Duration.ofSeconds(PASSWORD_WAIT_TIMEOUT_SECONDS).toMillis());

                } catch (Exception ex) {
                    logger.warn("PASSWORD_REQUEST flow failed: {}", ex.getMessage());
                }
            }

            readyState = true;
            logger.debug("Connected to session : {}", this.sessionId);

            Utils.saveSessionId(channelId, this.sessionId);
            return true;
        }
        return false;
    }

    /**
     * Disconnects the agent and performs any cleanup required.
     */

    public boolean disconnect() {

        if (!readyState) {
            logger.debug("Channel connection {} is already disconnected.", agentName);
        }

        boolean result = channelApi.disconnect(sessionId);

        if (result)
        {
            sessionId = null;
            readyState = false;
        }

        return result;

    }

    /**
     * Pull messages using an ReceiveConfig object.
     *
     * @param receiveConfig the receive config for receive operation
     * @return list of message events
     */
    public EventMessageResult receive(ReceiveConfig receiveConfig) {
        logger.debug("ConnectionChannel.receive: {}", receiveConfig);
        if (!isReady()) {
            return null;
        }

        return channelApi.receive(sessionId, receiveConfig);
    }

    /**
     * Gets all agents
     *
     * @return return the list of all agents
     */
    public List<AgentInfo> getActiveAgents() {

        logger.debug("ConnectionChannel.getActiveAgents: ");

        if (!isReady()) {
            return null;
        }

        return channelApi.getActiveAgents(sessionId);
    }

    /**
     * Starts an asynchronous polling routine that emits batches to the handler.
     *
     * @param messageHandler callback for message-arrival events
     */
    public void receiveAsync(AgentConnectionEventHandler messageHandler) {
        if (receiveThread == null) {
            receiveThread = new Thread(new RunnableReceive(this, messageHandler));
            receiveThread.start();
        } else {
            logger.debug("Asynchronous receive is already running.");
        }
    }

    /**
     * Starts an asynchronous polling routine that emits batches to the handler.
     *
     * @param messageHandler callback for message-arrival events
     */
    public void receiveAsync(AgentConnectionEventHandler messageHandler, ReceiveConfig initialReceiveConfig) {
        if (receiveThread == null) {
            receiveThread = new Thread(new RunnableReceive(this, messageHandler, initialReceiveConfig));
            receiveThread.start();
        } else {
            logger.debug("Asynchronous receive is already running.");
        }
    }

    /**
     * Send a message to a specific destination (agent or topic).
     *
     * @param msg         message body
     * @param destination destination routing value
     * @return true on success, false on failure
     */

    public boolean sendMessage(String msg, String destination) {
        return sendMessage(msg, "*", false);
    }

    /**
     * Broadcast or default-route send, depending on server defaults.
     *
     * @param msg message body
     * @return true on success, false otherwise
     */

    public boolean sendMessage(String msg) {
        return sendMessage(msg, "*", true);
    }

    /**
     * Send a message to destination, optionally marking destination as a regex filter.
     *
     * @param content       message body
     * @param destination   destination or filter
     * @param asFilterRegex whether destination is a regex to filter recipients
     * @return true on success, false otherwise
     */
    public boolean sendMessage(String content, String destination, boolean asFilterRegex) {

        if (!isReady()) {
            return false;
        }

        return channelApi.send(content, asFilterRegex ? destination : Pattern.quote(destination),
                sessionId);
    }

    public boolean isReady()
    {
        if (!readyState || sessionId == null) {
            logger.debug("Unable use channel operation, channel is not ready");
            return false;
        }
        else
        {
            return true;
        }
    }

    public static class RunnableReceive implements Runnable {

        private final AgentConnection channel;
        private final AgentConnectionEventHandler messageHandler;

        @Getter
        @Setter
        private ReceiveConfig offsetRange;

        public RunnableReceive(AgentConnection channel, AgentConnectionEventHandler messageHandler) {
            this(channel, messageHandler,  new ReceiveConfig(0L, 0L, DEFAULT_RECEIVE_LIMIT));
        }

        public RunnableReceive(AgentConnection channel, AgentConnectionEventHandler messageHandler,
                               ReceiveConfig initialReceiveConfig) {
            this.channel = channel;
            this.messageHandler = messageHandler;
            this.offsetRange = initialReceiveConfig;
        }

        @Override
        public void run() {
            while (channel.isReady()) {
                EventMessageResult messageEventResult = channel.receive(offsetRange);
                if (messageEventResult != null) {

                    List<EventMessage> events = messageEventResult.getEvents();
                    if (!events.isEmpty()) {
                        messageHandler.onMessageEvents(events);
                        // advance global and local offsets for next read
                        if (messageEventResult.getNextGlobalOffset() != null) {
                            this.offsetRange.setGlobalOffset(messageEventResult.getNextGlobalOffset());
                        }
                        if (messageEventResult.getNextLocalOffset() != null) {
                            this.offsetRange.setLocalOffset(messageEventResult.getNextLocalOffset());
                        }
                    }
                }
            }
        }
    }

    // Wait for a PASSWORD_REPLY addressed to this agent, decrypt it using the provided KeyPair
    // and set the channel secret via channelApi.setChannelSecret(secret). Returns true on success.
    private boolean waitForPasswordReply(Long globalOffset, Long localOffset, KeyPair kp, long timeoutMs) {
        final long deadline = System.currentTimeMillis() + timeoutMs;
        while (System.currentTimeMillis() < deadline) {
            EventMessageResult res = this.receive(new ReceiveConfig(globalOffset, localOffset, DEFAULT_RECEIVE_LIMIT));
            if (res != null && res.getEvents() != null) {
                for (EventMessage ev : res.getEvents()) {
                    if (ev.getDate() > this.connectionTime
                            && ev.getType() == EventMessage.EventType.PASSWORD_REPLY
                            && this.agentName.equals(ev.getTo()))
                    {
                        try {
                            byte[] cipher = java.util.Base64.getDecoder().decode(ev.getContent());
                            byte[] plain = EnvelopeUtil.rsaDecrypt(kp.getPrivate(), cipher);
                            String secret = new String(plain, StandardCharsets.UTF_8);
                            channelApi.setChannelSecret(secret);
                            return true;
                        } catch (Exception e) {
                            logger.warn("Failed to unwrap PASSWORD_REPLY: {}", e.getMessage());
                            // continue waiting for another reply until timeout
                        }
                    }
                }
                globalOffset = res.getNextGlobalOffset();
                localOffset = res.getNextLocalOffset();
            }
            try {
                Thread.sleep(400);
            } catch (InterruptedException ie) {
                Thread.currentThread().interrupt();
                break;
            }
        }
        return false;
    }
}
